### PR TITLE
Use promised-ssh to deploy projects instead of fabric

### LINKED ISSUE #36: Change to a JS-based deployment system
We don't really do much with fabric at the moment anyway, so having something JS-written do the same work would probably make it easier to test etc.

@relekang suggested https://github.com/pstadler/flightplan

### LINKED ISSUE #53: Support custom tasks
Add `tasks.deploy` as an option, which should be a list. If it is not set in a project it should use `config.DEFAULT_TASKS` which should be an env variable of the same name with `['make production']` as default. 

```json
"tasks": {
  "deploy": ["make production"],
  "restart": "forever restart $PWD/index.js",
"
```

### PR DESCRIPTION
**Pull request by relekang**
- [x] Throw out fabric
- [x] Support tasks in settings, fixes #53
- [x] Add promised-ssh in `Deployment.run`, fixes #36 
- [x] Rewrite tests, as spawn mocks no longer work.
- [x] Make sure promised-ssh hook up the event handlers for onStdout and onStderr.
- [x] Fix error handling with `.catch`

### CODE REVIEW

File: src/Deployment.js [hunk 1]
```diff
@@ -15,41 +16,52 @@ function Deployment(project, options) {
...
+    tasks.push('cd ' + this.projectConfig.path + ' && ' + task);
   }.bind(this));
 
-  proc.on('close', function(code) {
-    this.success = code === 0;
+  return ssh
+    .connect({
+      host: this.projectConfig.hostname,
+      username: this.projectConfig.user,
+      privateKey: fs.readFileSync(config.PATH_TO_PRIVATE_KEY),
```
Comment (ekmartin) [inline src/Deployment.js]: This should probably be done async, or read in at program launch in `./config` instead.

File: src/Deployment.js [hunk 2]
```diff
@@ -15,41 +16,52 @@ function Deployment(project, options) {
...
-  }.bind(this));
+  }.bind(this);
 
-  proc.stderr.on('data', function(data) {
+  var onStderr = function onStderr(data) {
     this.stderr += data;
     this.emit('stderr', data);
+  }.bind(this);
+
+  tasks = [];
```
Comment (ekmartin) [inline src/Deployment.js]: Missing `var` :)

File: src/Deployment.js [hunk 3]
```diff
@@ -15,41 +16,52 @@ function Deployment(project, options) {
...
-Deployment.prototype.deployScript = path.resolve('deploy.py');
-
-Deployment.prototype.run = function() {
-  var proc = childProcess.spawn(this.python, [this.deployScript, this.project]);
-  proc.stdout.setEncoding('utf8');
-  proc.stderr.setEncoding('utf8');
-
-  proc.stdout.on('data', function(data) {
+Deployment.prototype.run = function run() {
+  var onStdout = function onStdout(data) {
```
Comment (ekmartin) [inline src/Deployment.js]: Is the `var onStdout` necessary here? Instead of just `function onStdout`.

File: src/Deployment.js [hunk 4]
```diff
@@ -15,41 +16,52 @@ function Deployment(project, options) {
...
-    this.success = code === 0;
+  return ssh
+    .connect({
+      host: this.projectConfig.hostname,
+      username: this.projectConfig.user,
+      privateKey: fs.readFileSync(config.PATH_TO_PRIVATE_KEY),
+      onStdout: onStdout,
+      onStderr: onStderr,
+      debug: true
+    })
```
Comment (ekmartin) [inline src/Deployment.js]: You can use Bluebird's [bind](https://github.com/petkaantonov/bluebird/blob/master/API.md#binddynamic-thisarg---promise) on the Promise-chain so you don't have to do it manually on each `.then()`-function.

File: src/Deployment.js (hunk 2 above)
Comment (relekang) [inline src/Deployment.js]: Oops, fixed it :)

File: src/Deployment.js (hunk 4 above)
Comment (relekang) [inline src/Deployment.js]: Nice, I fixed it :)

File: src/Deployment.js (hunk 3 above)
Comment (relekang) [inline src/Deployment.js]: I am not sure if it really is necessary, but `var ... = function` creates the function on runtime not when the file is parsed, thus the function is created each time `run` is called which I thought was a good idea for eventlisteners.

File: Makefile [hunk 5]
```diff
@@ -19,14 +19,14 @@ all: src/public/main.css
 src/public/main.css: $(STYL)
 	$(STYLUS) > $@
 
-run: venv
+run:
 	ABAKUS_TOKEN=test HOOK_TOKEN=test SERVER_CONFIG_FILE=$(PWD)/example.json $(CHEWIE)
 
 test:
-	ABAKUS_TOKEN=test HOOK_TOKEN=test REDIS=true REDIS_DB=3 SERVER_CONFIG_FILE=$(PWD)/example.json $(BIN)/istanbul cover $(BIN)/_mocha test
+	PATH_TO_PRIVATE_KEY=$(PWD)/LICENSE BAKUS_TOKEN=test HOOK_TOKEN=test REDIS=true REDIS_DB=3 SERVER_CONFIG_FILE=$(PWD)/example.json $(BIN)/istanbul cover $(BIN)/_mocha test
```
Comment (ekmartin) [inline Makefile]: `BAKUS_TOKEN`? :smile:

File: src/Deployment.js [hunk 6]
```diff
@@ -15,41 +16,54 @@ function Deployment(project, options) {
...
+        onStderr: onStderr,
+        debug: true,
+        privateKey: key
+      });
+    })
+    .then(function(connection) {
+      return connection.exec(tasks);
+    })
+    .spread(function(code, stdout, stderr) {
+      this.success = code === 0;
```
Comment (ekmartin) [inline src/Deployment.js]: Is this necessary now, will it get here if code is `> 0`?

Comment (relekang) [inline src/Deployment.js]: Maybe, I have to do some changes to promised-ssh. I'll remove it in another PR later on.

File: src/controllers/hooks.js [hunk 7]
```diff
@@ -8,18 +8,17 @@ function deploy(project, res, next) {
...
+  deployment
+    .run()
+    .then(function() {
+      res.json({
+        status: 200,
+        output: deployment.stdout
+      });
+    })
+    .catch(function(err) {
       return next(err);
```
Comment (ekmartin) [inline src/controllers/hooks.js]: Can replace this with just `.catch(next)` :)

**Comment by ekmartin** [comment]
Looks great! :rocket: :racehorse:

File: src/controllers/hooks.js (hunk 7 above)
Comment (relekang) [inline src/controllers/hooks.js]: Good idea :)

Comment (relekang) [inline src/controllers/hooks.js]: Fixed it

**Pull request merged**